### PR TITLE
Allow void elements to be specified in data provider

### LINKED ISSUE #123: VOID Elements hardcoded for with non html langauges
So I'm making a custom extension that's parsing xml, and I'm building my extension on top of this html language service, this has worked great in most places because I can just disable the default data provider and roll my own.

The problem I'm having is with [fact.ts](https://github.com/microsoft/vscode-html-languageservice/blob/main/src/languageFacts/fact.ts) method `isVoidElement`. It has a hard coded list of elements that are void (for example ``) and this is a problem in my case because an `img` tag is not void in my case. I'd really like to be able to customize void elements. I'd even be happy to make a PR for the change, but I'd like to figure out the best way to implement it. Up till now I've done a bit of a hack to access the hard coded list and modify it for my use case, but that's causing some problems at build time for me now and I'd like to find a better way to solve my problem.

What makes sense to me is to modify the custom data providers to be able to specify this.
## Option 1 ##
Modify `HTMLDataV1` to have a string list identifying all the void tags. and add an api to `IHTMLDataProvider` to retrieve that list Like this: 
```ts
export interface HTMLDataV1 {
	version: 1 | 1.1;
	tags?: ITagData[];
	globalAttributes?: IAttributeData[];
	valueSets?: IValueSet[];
	voidTags?: string[]; // <===== new line
}

export interface IHTMLDataProvider {
	getId(): string;
	isApplicable(languageId: string): boolean;

	provideTags(): ITagData[];
	provideAttributes(tag: string): IAttributeData[];
	provideValues(tag: string, attribute: string): IValueData[];
	provideVoidTags(): string[]; // <===== new line
}
```
Then in various places where `isVoidElement` we could provide the list of void elements using the HTML Data, similar to what we're doing already for stuff like autocomplete data.

The up side to this change is that it works very similar to how we already customize the language service and would be easy for developer discovery. The downside is that it would require the most work of the changes I'm going to suggest.

## Option 2 ##
export the `VOID_ELEMENTS` list in `fact.js` as part of the API for the language service. I've hacked together a work around by doing this myself but it doesn't always play nice. This would allow the end developer to modify the list as they desire.

The up side is that this is a very simple change that doesn't require changing very much code at all. The downside is that it's hard for a dev to discovery, it's inconsistent with how things have been done so far. It's also a SINGLETON, so there's various problems with that as well.

## Option 3 ##
define a new language service option to specify VOID elements. 

This would be more difficult than option 2 but simpler than option 1 due to it being a single list that is always the same for a given instance of the service, it would also be decent for developer discovery as it's part of the options API exposed when consuming the service. The downside is that it's not at all consistent with how custom data has been handled in the past.

Please let me know what you think, and what direction would be the best to take, then I can start work on a PR.

### PR DESCRIPTION
**Pull request by hahn-kev**
This fixes #123 

Please note, this PR drops support for checking void elements `keygen` and `menuitem` these are long ago depreciated and not used it any tests. If you need these you can add them back in via a custom data provider and set `"void": true`.

### CODE REVIEW

**Comment by aeschli** [comment]
I suggest to make `isVoidElement` a member of HTMLDataManager.

HTMLCompletion already gets a HTMLDataManager in the constructor.

I'd turn folding and the parser also a class that gets the HTMLCompletion as context.
Same for the selection range. It could get a parser as context

File: src/services/htmlSelectionRange.ts [hunk 1]
```diff
@@ -4,14 +4,21 @@
...
+export class HTMLSelectionRange {
 
-	function getSelectionRange(position: Position): SelectionRange {
-		const applicableRanges = getApplicableRanges(document, position);
-		let prev : [number, number] | undefined = undefined;
+	constructor(private htmlParser: HTMLParser) {
+	}
+
+	public getSelectionRanges(document: TextDocument, positions: Position[]): SelectionRange[] {
+		const htmlDocument = this.htmlParser.parseDocument(document);
```
Comment (hahn-kev) [inline src/services/htmlSelectionRange.ts:16]: this is a slight tweak to parse the document here, it used to be parsed inside `getApplicableRanges` which would happen for each position, now it only parses the document once.

**Comment by hahn-kev** [comment]
ok @aeschli I've created classes for those 3 functions. The formatting makes it look like there are more changes than there really is, but I had to indent all the functions since they are all inside classes now.

**Review by aeschli — APPROVED** [review]

**Review by dbaeumer — APPROVED** [review]

**Comment by aeschli** [comment]
Thanks a lot @hahn-kev, nice PR!

**Comment by hahn-kev** [comment]
You're welcome, happy to contribute

**Pull request merged**